### PR TITLE
LOR for all element types [lor-all-elements]

### PR DESCRIPTION
**Pull request by pazner**
This PR enables the creation of LOR-type meshes for all element types (including mixed meshes).

The `L2ProjectionGridTransfer` is updated so that the `lor-transfer` miniapp works with these meshes too.

star
star_lor

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2014](https://github.com/mfem/mfem/pull/2014) | @pazner | @tzanio | @tzanio + @barker29 + @jakubcerveny | 01/25/21 | 02/09/21 | 02/14/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
So cool!

**Comment by tzanio** [comment]
Please mention in `CHANGELOG`

**Review by tzanio — APPROVED** [review]
Looks great, thanks @pazner!

File: fem/fespace.cpp [hunk 1]
```diff
@@ -2750,74 +2750,95 @@ L2ProjectionGridTransfer::L2Projection::L2Projection(
...
    for (int ilor=0; ilor<nel_lor; ++ilor)
    {
       int iho = cf_tr.embeddings[ilor].parent;
       ho2lor.AddConnection(iho, ilor);
    }
    ho2lor.ShiftUpI();
 
+   offsets.SetSize(nel_ho+1);
+   offsets[0] = 0;
+   for (int iho=0; iho<nel_ho; ++iho)
```
Comment (jakubcerveny) [inline fem/fespace.cpp:2783]: I can't find anything substantial so I'm nitpicking: the whitespace here and there could be unified, as in `for (int iho = 0; iho < nel_ho; ++iho)`, it's less crammed and generally preferred this way in MFEM...

**Review by jakubcerveny — APPROVED** [review]
This looks good, I tried the LOR miniapp and it seems to work with any mesh now (e.g., `toroid-wedge.mesh`)
toroid-wedge

**Comment by jakubcerveny** [comment]
Some suggestions not really related to this PR but that maybe could be added:

- The description of `lor-transfer.cpp` talks about `FiniteElementSpace::GetTransferOperator` and `FiniteElementSpace::GetReverseTransferOperator`, but these are never used by the miniapp, is the description outdated?
- `class GridTransfer` and descendants would perhaps deserve their own module (`gridtransfer.?pp`?), ~700 lines is already enough and also this stuff is not really that close to `FiniteElementSpace`?

File: fem/fespace.hpp [hunk 2]
```diff
@@ -947,12 +947,10 @@ class L2ProjectionGridTransfer : public GridTransfer
       const FiniteElementSpace &fes_ho;
       const FiniteElementSpace &fes_lor;
 
-      int ndof_lor, ndof_ho, nref;
+      mutable Array<double> R, P;
+      Array<int> offsets;
```
Comment (barker29) [inline fem/fespace.hpp:951]: I realize this is private, but I think brief documentation of `offsets` would make the code easier to maintain.

**Review by barker29 — APPROVED** [review]
Like Jakub I'm having trouble finding anything substantive to say, this is very elegantly done and I'm impressed that this functionality can be added in ~50 net lines of code.

I agree with Jakub's two points regarding `lor-transfer.cpp` documentation and putting `GridTransfer` in a separate file. In addition, I suggest adding a new sample run to the top of `lor-transfer.cpp` that uses a tri or tet mesh.

File: fem/fespace.cpp (hunk 1 above)
Comment (pazner) [reply fem/fespace.cpp:2783]: Sure, I changed these.

File: fem/fespace.hpp (hunk 2 above)
Comment (pazner) [reply fem/fespace.hpp:956]: Good idea, I added a comment.

**Comment by pazner** [comment]
> * The description of `lor-transfer.cpp` talks about `FiniteElementSpace::GetTransferOperator` and `FiniteElementSpace::GetReverseTransferOperator`, but these are never used by the miniapp, is the description outdated?

Good catch, I updated the comment.

**Comment by pazner** [comment]
> Like Jakub I'm having trouble finding anything substantive to say, this is very elegantly done and I'm impressed that this functionality can be added in ~50 net lines of code.

Thanks :)

> I agree with Jakub's two points regarding `lor-transfer.cpp` documentation and putting `GridTransfer` in a separate file. In addition, I suggest adding a new sample run to the top of `lor-transfer.cpp` that uses a tri or tet mesh.

Good idea, I added a mixed mesh sample run.

Regarding moving the transfer operators to their own file: I agree completely, but I think that should probably be a PR on its own.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**